File: Two pointer/167.Two_Sum_II_Input_Array_Is_Sorted/Two_Sum_II_Input_Array_Is_Sorted.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        vector<int> ans;
        
        int start=0;
        int end=numbers.size()-1;

        while(start<=end){
            if(numbers[start]+numbers[end]==target){  //If the sum is equal to the target, return the indices
                ans.push_back(start+1); 
                ans.push_back(end+1);
                return ans;
            }
            else if(numbers[start]+numbers[end]>target){ //If the sum is greater than the target, decrement the end pointer
                end--;
            }else{
                start++;
            }
        }
        return ans;
    }
};
int main() {
    Solution sol;
    vector<int> numbers = {2, 7, 11, 15};
    int target = 9;
    vector<int> result = sol.twoSum(numbers, target);

    for (int i : result) {
        cout << i << " ";
    }
    cout << endl;

    return 0;
}

File: Two pointer/167.Two_Sum_II_Input_Array_Is_Sorted/Two_Sum_II_Input_Array_Is_Sorted.scala
object TwoSumII {
    def twoSum(numbers: Array[Int], target: Int): Array[Int] = {
        var left = 0
        var right = numbers.length - 1

        while (left < right) {
            val sum = numbers(left) + numbers(right)
            if (sum == target) {
                return Array(left + 1, right + 1)
            } else if (sum < target) {
                left += 1
            } else {
                right -= 1
            }
        }
        Array()
    }

    def main(args: Array[String]): Unit = {
        val numbers = Array(2, 7, 11, 15)
        val target = 9
        val result = twoSum(numbers, target)
        println(s"Indices: ${result.mkString(", ")}")
    }
}